solana/bridge_token_factory/programs/bridge_token_factory/src/instructions/user/register_mint.rs

```diff
@@ -0,0 +1,147 @@
+
+use super::MetadataPayload;
+use crate::{
+    constants::{AUTHORITY_SEED, CONFIG_SEED, MESSAGE_SEED, VAULT_SEED},
+    state::config::Config,
+};
+use anchor_spl::metadata::ID as MetaplexID;
+
+#[derive(Accounts)]
+pub struct RegisterMint<'info> {
+    #[account(
+        seeds = [CONFIG_SEED],
+        bump = config.bumps.config,
+    )]
+    pub config: Account<'info, Config>,
+    /// CHECK: PDA
+    #[account(
+        seeds = [AUTHORITY_SEED],
+        bump = config.bumps.authority,
+    )]
+    pub authority: UncheckedAccount<'info>,
+
+    #[account(
+        constraint = !mint.mint_authority.contains(authority.key),
+    )]
+    pub mint: Account<'info, Mint>,
+
+    #[account(
+        seeds = [
+            b"metadata",
+            MetaplexID.as_ref(),
+            &mint.key().to_bytes(),
+        ],
+        bump,
+        seeds::program = MetaplexID,
+    )]
+    pub metadata: Account<'info, MetadataAccount>,
+
+    #[account(
+        init,
+        payer = payer,
+        token::mint = mint,
+        token::authority = authority,
+        seeds = [
+            VAULT_SEED,
+            mint.key().as_ref(),
+        ],
+        bump,
+    )]
+    pub vault: Account<'info, TokenAccount>,
+
+    /// Wormhole bridge data. [`wormhole::post_message`] requires this account
+    /// be mutable.
+    #[account(
+        address = config.wormhole.bridge,
+    )]
+    pub wormhole_bridge: Account<'info, BridgeData>,
+
+    /// Wormhole fee collector. [`wormhole::post_message`] requires this
+    /// account be mutable.
+    #[account(
+        mut,
+        address = config.wormhole.fee_collector
+    )]
+    pub wormhole_fee_collector: Account<'info, FeeCollector>,
+
+    /// Emitter's sequence account. [`wormhole::post_message`] requires this
+    /// account be mutable.
+    #[account(
+        mut,
+        address = config.wormhole.sequence
+    )]
+    pub wormhole_sequence: Account<'info, SequenceTracker>,
+
+    /// CHECK: Wormhole Message. [`wormhole::post_message`] requires this
+    /// account be mutable.
+    #[account(
+        mut,
+        seeds = [
+            MESSAGE_SEED,
+            &wormhole_sequence.next_value().to_le_bytes()[..]
+        ],
+        bump,
+    )]
+    pub wormhole_message: UncheckedAccount<'info>,
+
+    #[account(mut)]
+    pub payer: Signer<'info>,
+
+    pub clock: Sysvar<'info, Clock>,
+    pub rent: Sysvar<'info, Rent>,
+
+    pub system_program: Program<'info, System>,
+    pub token_program: Program<'info, Token>,
+    pub associated_token_program: Program<'info, AssociatedToken>,
+    pub wormhole_program: Program<'info, Wormhole>,
+}
+
+impl<'info> RegisterMint<'info> {
+    pub fn process(&mut self, wormhole_message_bump: u8) -> Result<()> {
+        // TODO: correct message payload
+        let payload = MetadataPayload {
+            token: self.mint.key().to_string(),
+            name: self.metadata.name.clone(),
```

Comment:
Not sure. It is possible to put it together, but we need to make metaplex metadata account optional. If we want to support only 2 metadata protocols (metaplex and native token22) it will be ok using Interface. Basically, the main reason of separation is different metadata standard not token version